lib/websocket/websocket-handler.ts

```diff
@@ -0,0 +1,238 @@
+    var receivedSubscriptions = new SubscriptionStore<Subscription>();
+
+    // Clean up sockets for cases where the underlying session terminated unexpectedly
+    socket.blpSession.once('SessionTerminated', (): void => {
+        if (socket.isConnected()) {
+            socket.log.debug('blpSession terminated unexpectedly. Terminate socket.');
+            socket.emit('err', { message: 'blpSession terminated unexpectedly.' });
+            socket.disconnect();
+        }
+    });
+
+    // Subscribe
+    socket.on('subscribe', (data: Object[]): void => {
+        socket.log.info('Subscribe request received');
+        // Log body if configured
+        if (conf.get('logging.reqBody')) {
+            socket.log.debug({body: data}, 'Subscribe body.');
+        }
+
+        // Validate input options
+        var subscriptions : Subscription[] = [];
+        if (!data.length) {
+            socket.log.debug('No valid subscriptions found.');
+            socket.emit('err', { message: 'No valid subscriptions found.' });
+            return;
+        }
+        data.forEach((s: any): void => {
+            // Check if all requests are valid
+            // The Subscribe request will proceed only if all subscriptions are valid
+            if (!_.has(s, 'correlationId') || !_.isNumber(s.correlationId)
+                || !_.has(s, 'security') || !_.isString(s.security)
+                || !_.has(s, 'fields') || !_.isArray(s.fields)) {
+                socket.log.debug('Invalid subscription option.');
+                socket.emit('err', { message: 'Invalid subscription option.' });
+                return;
+            }
+            if (receivedSubscriptions.has(s.correlationId)) {
+                socket.log.debug('Correlation Id already exists.');
+                socket.emit('err', { message: 'Correlation Id already exists.' });
+                return;
+            }
+
+            var sub = new Subscription(s.correlationId,
+                                            s.security,
+                                            s.fields,
+                                            s.options);
+            subscriptions.push(sub);
+            receivedSubscriptions.add(sub);
+
+            // Add event listener for each subscription
+            sub.on('data', (data : any) : void => {
+// For debug purpose
+var ts = process.hrtime();
+socket.log.debug({data: {cid: sub.correlationId, time: ts}}, 'Data received');
+data['DEBUG_TIME'] = ts[0] + '.' + ts[1];
+
+                // Emit the data
+                // TODO: Acknowledgement callback function?
+                socket.emit('data', {'correlationId' : sub.correlationId,
+                                     'data' : data});
+                socket.log.info('Data sent');
+            });
+        });
+
+        // Subscribe user request through blpapi-wrapper
+        socket.blpSession.subscribe(subscriptions)
+        .then((): void => {
```

Comment:
I should probably just remove this part of the code since it is attaching stuff to the actual data object sent back.